lib/relay/client/dynamic.go

```diff
@@ -150,15 +151,26 @@ type dynamicAnnouncement struct {
 func relayAddressesOrder(input []string) []string {
 	buckets := make(map[int][]string)
 
+	wg := sync.NewWaitGroup()
+	wg.Add(len(input))
+	results := make(chan urlWithLatency, len(input))
 	for _, relay := range input {
-		latency, err := osutil.GetLatencyForURL(relay)
-		if err != nil {
-			latency = time.Hour
-		}
+		go func(irelay string) {
+			latency, err := osutil.GetLatencyForURL(irelay)
+			if err != nil {
+				latency = time.Hour
+			}
+			results <- urlWithLatency{irelay, latency}
+			wg.Done()
+		}(relay)
+	}
 
-		id := int(latency/time.Millisecond) / 50
+	wg.Wait()
+	close(results)
```

Comment:
I think this doesn't work: The loop below won't read anything from results, as range aborts if the looped over channel is closed. Personally I wouldn't buffer the channel and put the two lines above into a goroutine.

Reply:
https://play.golang.org/p/4cGK8dy18Dh

Reply:
Ah right, so the close kind of gets appended to the buffered elements in the channel. I guess that makes sense. Thanks for teaching me :)